### PR TITLE
Feature/boss manager

### PR DESCRIPTION
**Pull request by Minoord**

### CODE REVIEW

File: Assets/Art/Textures/UI/Healthbar/Base-no_skulls.png
Comment (swzwij) [inline Assets/Art/Textures/UI/Healthbar/Base-no_skulls.png:1]: T_BaseNoSkulls

File: Assets/Art/Textures/UI/Healthbar/Skulls.png
Comment (swzwij) [inline Assets/Art/Textures/UI/Healthbar/Skulls.png:1]: T_Skulls

File: Assets/Scripts/Pathfinding/PathFollower.cs [hunk 1]
```diff
@@ -32,6 +32,8 @@ public class PathFollower : NetworkBehaviour
         protected Coroutine p_followPathCoroutine;
         protected Coroutine p_waitForClearance;
         protected LineRenderer p_pathRenderer;
+
+        public Pathfinder MyPathFinder { set { p_pathfinder = value; } }
```
Comment (swzwij) [inline Assets/Scripts/Pathfinding/PathFollower.cs]: Can this be renamed to just `PathFinder` ?

File: Assets/Scripts/PoolSystem/PoolSystem.cs [hunk 2]
```diff
@@ -100,9 +103,12 @@ private IEnumerator RespawnTimer(int respawnTime, string objectName, Transform p
...
         {
-            if (parentTransform == null)
-                SceneManager.MoveGameObjectToScene(movingObject, SceneManager.GetActiveScene());
-            else
+/*            if (parentTransform == null)
+            {
+                movingObject.transform.parent = 
+                //SceneManager.MoveGameObjectToScene(movingObject, SceneManager.GetActiveScene());
+            }
+            else*/
```
Comment (swzwij) [inline Assets/Scripts/PoolSystem/PoolSystem.cs]: Please remove commented code

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+using Examen.Networking;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class EnemySpawnPoints : MonoBehaviour
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs]: Missing namespace

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+{
+   private Transform _spawnpoint;
+   private List<Transform> _waypoints = new();
+
+    public Transform Spawnpoint {  get { return _spawnpoint; } }
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs]:
```suggestion
    public Transform Spawnpoint => _spawnpoint;
```

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    public List<Transform> Waypoints {  get { return _waypoints; } }
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs]:
```suggestion
    public List<Transform> Waypoints => _waypoints;
```

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Examen.Spawning.BossSpawning
+{
+    public class EnemySpawner : NetworkedSingletonInstance<EnemySpawner>
+    {
+        [Header("Boss Spawning")]
+        [SerializeField] private int _enemyDownTimer = 60;
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs]: Please rename to either interval, or cooldown

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        private bool _hasConnected;
+        private Dictionary<EnemyHealthBar, float> _enemiesHealth = new();
+
+        private void Start()
+        {
+            ServerInstance.Instance.OnServerStarted += SpawnEnemy;
+            _spawner.OnSpawned += SendEnemyInfoOnConnection;
+        }
+
+        public void DespawnEnemy(EnemyHealthBar enemyHealthBar)
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs:39]: Missing summary

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+    private void Start() => SetWaypoints();
+
+    public void SetWaypoints()
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawnPoints.cs]: Missing summary

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        {
+            PoolSystem.Instance.DespawnObject(enemyHealthBar.gameObject.name, enemyHealthBar.gameObject);
+            _enemiesHealth.Remove(enemyHealthBar);
+        }
+
+        private void SendEnemyInfoOnConnection(NetworkObject obj) => ReceiveEnemyInfoOnConnection(_enemiesHealth);
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs]: Please fully spell out parameter names

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+        [ObserversRpc]
+        private void ReceiveEnemyInfoOnConnection(Dictionary<EnemyHealthBar, float> currentActiveEnemySliders)
+        {
+            if (_hasConnected)
+                return;
+
+            foreach (KeyValuePair<EnemyHealthBar, float> sliders in currentActiveEnemySliders)
+                sliders.Key.ClientInitialize(sliders.Value);
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs:54]:
```suggestion
            foreach (EnemyHealthBar slider in currentActiveEnemySliders.Keys)
                slider.ClientInitialize(sliders.Value);
```

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            _hasConnected = true;
+        }
+
+        [Server]
+        private void SpawnEnemy()
+        {
+            int randomSpawnPointNumber = Random.Range(0, _enemySpawnPoints.Count);
+            int randomEnemyPrefabNumber = Random.Range(0, _enemyPrefabs.Count);
+
+            GameObject enemy = PoolSystem.Instance.SpawnObject(_enemyPrefabs[randomEnemyPrefabNumber].name, _enemyPrefabs[randomEnemyPrefabNumber].gameObject);
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs]: Line exceeds the 120 character limit

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+            AddSlider(enemy);
+        }
+
+        [ObserversRpc]
+        private void EnableEnemy(GameObject enemy) => enemy.SetActive(true);
+
+
+        private IEnumerator DelayedPathStart(GameObject enemy, int randomSpawnPointNumber)
+        {
+            yield return new WaitForSeconds(0.33f);
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs]: Does this need to be hard coded in? if so either use a variable or a constant

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+            healthBar.EnemyHealthData = enemyHealth;
+            healthBar.ServerInitialize();
+            ReceiveEnemyInfoOnSpawn(healthBar, enemyHealth.MaxHealth);
+
+            _enemiesHealth.Add(healthBar, enemyHealth.Health);
+        }
+
+        [ObserversRpc]
+        private void ReceiveEnemyInfoOnSpawn(EnemyHealthBar healthbar, float healthAmount) => healthbar.ClientInitialize(healthAmount);
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs]: Line exceeds the 120 character limit

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+
+        private void DespawnEnemy()
+        {
+            foreach (KeyValuePair<EnemyHealthBar, float> slider in _enemiesHealth)
+                PoolSystem.Instance.DespawnObject(slider.Key.name, slider.Key.gameObject);
```
Comment (swzwij) [inline Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs]: Same as the other for each loop comment

File: Assets/Scripts/UI/EnemyHealthBar.cs [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+using FishNet.Object;
+using MarkUlrich.Health;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Examen.UI
+{
+    public class EnemyHealthBar : NetworkBehaviour
+    {
+        [SerializeField] private bool _damage;
```
Comment (swzwij) [inline Assets/Scripts/UI/EnemyHealthBar.cs]: Remove

File: Assets/Scripts/UI/EnemyHealthBar.cs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        [SerializeField] private Slider _healthBar;
+
+        public HealthData EnemyHealthData { get; set; }
+        public Slider HealthSlider => _healthBar;
```
Comment (swzwij) [inline Assets/Scripts/UI/EnemyHealthBar.cs]:
```suggestion
        public Slider HealthBar => _healthBar;
```
please keep the variable names the same

File: Assets/Scripts/UI/EnemyHealthBar.cs [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+        private void Update()
+        {
+            if (!IsServer)
+                _healthBar.transform.parent.rotation = Quaternion.LookRotation(Vector3.back);
+
+            if (_damage) EnemyHealthData.TakeDamage(1);
```
Comment (swzwij) [inline Assets/Scripts/UI/EnemyHealthBar.cs]:
```suggestion
            if (_damage) 
                EnemyHealthData.TakeDamage(1);
```

File: Assets/Scripts/Pathfinding/WaypointFollower.cs [hunk 18]
```diff
@@ -11,29 +11,19 @@ public class WaypointFollower : PathFollower
...
-
-            _waypoints.Reverse();
-
-            if (_waypoints.Count == 0)
-                return;
+            GenerateCompletePath();
+        }
 
+        [Server]
+        public void ResetWaypointIndex()
```
Comment (swzwij) [inline Assets/Scripts/Pathfinding/WaypointFollower.cs:27]: Missing summary

**Review by swzwij — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs (hunk 10 above)
Comment (Minoord) [reply Assets/Scripts/Spawning/BossSpawning/EnemySpawner.cs:54]: I can't do this, because I both need the key and value

**Review by swzwij — APPROVED** [review]

**Pull request merged**